chainstate/src/rpc/types/input.rs

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2024 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common::{
+    address::AddressError,
+    chain::{ChainConfig, GenBlock, OutPointSourceId, Transaction, TxInput},
+    primitives::Id,
+};
+
+use super::account::{RpcAccountCommand, RpcAccountSpending};
+
+#[derive(Debug, Clone, serde::Serialize)]
+pub enum RpcTxInput {
+    Utxo {
+        source_id: RpcOutPointSourceId,
+        index: u32,
+    },
+    Account {
+        nonce: u64,
+        account: RpcAccountSpending,
+    },
+    AccountCommand {
+        nonce: u64,
+        command: RpcAccountCommand,
+    },
+}
+
+impl RpcTxInput {
+    pub fn new(chain_config: &ChainConfig, input: &TxInput) -> Result<Self, AddressError> {
+        let result = match input {
+            TxInput::Utxo(outpoint) => match outpoint.source_id() {
+                OutPointSourceId::Transaction(id) => RpcTxInput::Utxo {
+                    source_id: RpcOutPointSourceId::Transaction { tx_id: id },
+                    index: outpoint.output_index(),
+                },
+                OutPointSourceId::BlockReward(id) => RpcTxInput::Utxo {
+                    source_id: RpcOutPointSourceId::BlockReward { block_id: id },
+                    index: outpoint.output_index(),
+                },
+            },
+            TxInput::Account(outpoint) => RpcTxInput::Account {
+                nonce: outpoint.nonce().value(),
+                account: RpcAccountSpending::new(chain_config, outpoint.account().clone())?,
+            },
+            TxInput::AccountCommand(nonce, command) => RpcTxInput::AccountCommand {
+                nonce: nonce.value(),
+                command: RpcAccountCommand::new(chain_config, command)?,
+            },
+        };
+        Ok(result)
+    }
+}
+
+#[derive(Debug, Clone, serde::Serialize)]
+pub enum RpcOutPointSourceId {
+    Transaction { tx_id: Id<Transaction> },
+    BlockReward { block_id: Id<GenBlock> },
+}
```

Comment:
This is still bad.

You should never use an enum for json this way. This is because it will not make the user understand the different possibilities.

Again, please do it like it's done for the API server:

```
    match utxo.source_id() {
        OutPointSourceId::Transaction(tx_id) => {
            json!({
                "source_type": "Transaction",
                "source_id": tx_id.to_hash().encode_hex::<String>(),
                "index": utxo.output_index(),
            })
        }
        OutPointSourceId::BlockReward(block_id) => {
            json!({
                "source_type": "BlockReward",
                "source_id": block_id.to_hash().encode_hex::<String>(),
                "index": utxo.output_index(),
            })
        }
    }
```

Reply:
done, take a look

Reply:
example for TxInput:
```
'inputs': [{'input_type': 'Utxo', 'value': {'Utxo': {'index': 0, 'source_id': {'source_type': 'BlockReward', 'value': {'BlockReward': 'efcdf78d8c51794c01471e473bbfbd51db60559a63ece2a7b29114357672e845'}}}}}]
```